### PR TITLE
Implement basic branch and call opcodes

### LINKED ISSUE #52: Implement basic branch opcodes
The following opcodes for branching within a function should be implemented:

- `br`
- `br_if`
- `br_unless`

All WebAssembly opcodes for control flow within a function except `br_table` are lowered down to these 3 opcodes before being passed to the interpreter, so none of the WebAssembly structured control flow opcodes need to be considered.

### PR DESCRIPTION
**Pull request by Leonardo2718**
Implement and test the following opcodes:

* `br`
* `br_if`
* `br_unless`
* `call`

Close #52

### CODE REVIEW

File: test/jit/br.txt
Comment (aviansie-ben) [inline test/jit/br.txt]: I don't think these label-related tests are necessary. Labels are resolved in the `wat2wasm` tool and don't actually exist at all in the WebAssembly binary, so the interpreter and JIT never get a chance to see them.

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: These `th->Run` calls are completely broken for several reasons and will need serious changes to work correctly:

1. The value you passed in means that only a maximum of 1000 instructions will ever be executed by the new interpreter frame you spawn. You need to check the return value and re-call the interpreter if it's not done executing yet. The proper method of calling the interpreter can be seen here (although the tracing can probably be ignored): https://github.com/wasmjit-omr/wasmjit-omr/blob/deabf976619f07946f38c8ffe3c7301d1330debb/src/interp.cc#L3180-L3199
2. The interpreter will return `interp::Result::Returned` when it has completed. `CHECK_TRAP_IN_HELPER` will interpret this as a trap, causing the JITted function to exit prematurely to the caller, which will exhibit the same behaviour and thus kill the enclosing interpreter completely.
3. The new interpreter frame you spawn here doesn't know that the call frame you just pushed should not be executed in the interpreter. Thus, when it encounters the `ret` instruction in the callee, it will start interpreting instructions from call frames that it shouldn't, including the JITted function (at least, it would if the prior issue involving pushing the wrong address on the call stack were resolved).

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: You've pushed the wrong address onto the call stack! You need to push the address of the instruction to return to, not the address of the callee.

File: test/jit/call.txt
Comment (aviansie-ben) [inline test/jit/call.txt:2]: We're missing tests involving a JITted function calling into an interpreted function. It might be a bit tough to fit that into the existing test framework (since we need a way to ensure a function isn't JITted even though `--trap-on-failed-comp` is enabled), but I think such tests would be invaluable, since that situation seems to have quite a few gotchas.

**Review by aviansie-ben — CHANGES_REQUESTED** [review]
The implementation of the `call` opcode when calling an interpreted function from a JITted function is completely broken and needs major changes.

File: test/jit/call.txt
Comment (Leonardo2718) [reply test/jit/call.txt:2]: Agreed.

I haven't thought of a clean way of implementing JIT-to-interpreter call tests yet, given the test framework we have. Currently, there is no way of blacklisting functions so that they don't get JIT compiled. Since the compilation threshold is 1 for all JIT tests, the called function is necessarily going to be JIT compiled on its first call. Increasing the threshold might be a start but the test would have to be carefully crafted so that the caller is called multiple times (so it can get JIT compiled) and without having it call the callee function (so that it won't get JIT compiled).

File: test/jit/br.txt
Comment (Leonardo2718) [reply test/jit/br.txt]: Fair enough.

File: src/jit/function-builder.cc
Comment (Leonardo2718) [reply src/jit/function-builder.cc]: Yikes! That is not what that was supposed to do. Glad you caught that!

File: src/jit/function-builder.cc
Comment (Leonardo2718) [reply src/jit/function-builder.cc]: OK, I guess forgot to add the WIP tag to this PR :sweat_smile: .

I was aware of 1 and 3, but somehow managed to miss 2. I'll keep working at it :slightly_smiling_face: .

**Comment by Leonardo2718** [comment]
@aviansie-ben I've removed the "branch to label" tests and improved the call implementation. I still don't have a good way of testing JIT-to-interpreter dispatch though.

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: This is still pushing the wrong address because the thread's `pc_` value is stale at this point. Not only does it not get updated in the JITted function, the outer interpreter doesn't even save its value within interpreted code until it returns.

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: The `th->set_pc` is unnecessary here, since the value in there will never be used.

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: This looks like a typo: I think you meant to use `th->call_stack_top_` here based on your other code.

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: You already set the PC correctly at the start of `call_interp`, so this call is redundant.

**Review by aviansie-ben — CHANGES_REQUESTED** [review]

**Comment by aviansie-ben** [comment]
As for the tests for JIT-to-interpreter dispatch: I haven't thought of any good way to test that either. Maybe it's best if we ignore that for the time being and add an issue about it after merging...

File: src/jit/function-builder.cc
Comment (Leonardo2718) [reply src/jit/function-builder.cc]: *Sigh...* Not sure why I keep getting this one wrong :disappointed: .

File: src/jit/function-builder.cc
Comment (Leonardo2718) [reply src/jit/function-builder.cc]: Yep, definitely a typo.

File: src/jit/function-builder.cc
Comment (Leonardo2718) [reply src/jit/function-builder.cc]: Yay, more typos!

File: src/jit/function-builder.cc
Comment (Leonardo2718) [reply src/jit/function-builder.cc]: Initially, I wanted that call to be there to "perfectly-mimic" the behaviour of the interpreter. However, since we're already not doing that in other places, I'll get rid of it.

**Comment by Leonardo2718** [comment]
@aviansie-ben OK, it should be getting close now.

I agree that we should probably just open an issue about the tests so we can move the PR forward. For the recorded, I have done some manual testing locally by running variations of:

```wasm
  (func $i32_id_callee (param i32) (result i32)
    get_local 0)

  (func $i32_id_caller_if (param i32) (param i32) (result i32)
    get_local 1
    if
        get_local 0
        call $i32_id_callee
        return
    end
    get_local 0)

  (func (export "test_i32_id_call_0") (result i32)
    i32.const 3
    i32.const 0
    call $i32_id_caller_if
    i32.const 0
    call $i32_id_caller_if
    i32.const 1
    call $i32_id_caller_if
    )
```

and setting the compilation threshold to 2.

File: test/jit/br.txt
Comment (aviansie-ben) [inline test/jit/br.txt]: It looks like you accidentally removed this function when you removed some of the other tests.

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: For consistency with existing code, please add a space between `)` and `{`

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: No need to call `thread_->GetIstream()` here, the correct value is already in the `istream` parameter (same problem in `InterpBrUnless`).

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: For consistency, please add a space between `)` and `{` and put the `const` specifier before the type name (same problem in `InterpBrUnless`).

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: Indentation here is incorrect (same problem in `InterpBrUnless`).

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: This `)` should be on the previous line.

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: There's an unnecessary space between closing parentheses here.

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: There's an unnecessary space between closing parentheses here.

**Review by aviansie-ben — CHANGES_REQUESTED** [review]
The call helper looks good to me now. Remaining issues are mostly related to code style.

File: src/jit/function-builder.cc
Comment (aviansie-ben) [inline src/jit/function-builder.cc]: The `const` specifier here (and in `InterpBrUnless`) is still after the type name. Please move it to before the type name for consistency with other WABT code.

**Review by aviansie-ben — CHANGES_REQUESTED** [review]

**Review by aviansie-ben — APPROVED** [review]
LGTM. Will merge once checks pass.

**Pull request merged**
